docs/broker/filtering.md

```diff
@@ -0,0 +1,360 @@
+    that can be cached in compiled form.
+*   **Expressions are serializable and composable.** These properties are useful
+    in implementing upstream filter propagation. For example, a Broker could
+    advertise a single CEL expression composed of the union of its Triggers'
+    filters, allowing upstream sources to filter events before publishing to the
+    Broker.
+
+CEL is
+[memory-safe, side-effect-free, and terminating](https://github.com/google/cel-spec/blob/master/doc/langdef.md#overview),
+making it safe and secure to embed in a multi-tenant dispatcher process.
+
+## Examples
+
+### Single attribute exact match
+
+Specified with the SourceAndType field:
+
+```yaml
+spec:
+  filter:
+    sourceAndType:
+      type: com.github.pull.create
+```
+
+Specified with a CEL expression:
+
+```yaml
+spec:
+  filter:
+    cel:
+      expression: ce.type == "com.github.pull.create"
+```
+
+### Multiple attributes exact match
+
+Specified with the SourceAndType field:
+
+```yaml
+spec:
+  filter:
+    sourceAndType:
+      type: com.github.pull.create
+      source: https://github.com/knative/eventing/pulls/123
+```
+
+Specified with a CEL expression:
+
+_The `>` syntax is a standard yaml multiline string._
+
+```yaml
+spec:
+  filter:
+    cel:
+      expression: >
+        ce.type == "com.github.pull.create" &&
+        ce.source == "https://github.com/knative/eventing/pulls/123"
+```
+
+### Prefix match on source
+
+Cannot be specified with the SourceAndType field.
+
+CEL doesn't support prefix match directly, but prefix match can be accomplished
+with a regular expression:
+
+```yaml
+spec:
+  filter:
+    cel:
+      expression: ce.source.matches("^https://github.com")
+```
+
+### Complex boolean expression
+
+Cannot be specified with the SourceAndType field.
+
+_The `>` syntax is a standard yaml multiline string._
+
+```yaml
+spec:
+  filter:
+    cel:
+      expression: >
+        ce.type == "com.github.pull.create" ||
+        (ce.type == "com.github.issue.create" && ce.source.matches("proposals")
+
+```
+
+### Parsed extensions
+
+Cannot be specified with the SourceAndType field.
+
+_This example assumes the repository name is available as a CloudEvents
+extension with name `repository`._
+
+_The `>` syntax is a standard yaml multiline string._
+
+```yaml
+spec:
+  filter:
+    cel:
+      parseExtensions: true
+      expression: >
+        ce.type == "com.github.pull.create" ||
+         (ce.type == "com.github.issue.create" &&
+          ce.repository == "proposals")
+
+```
+
+### Parsed data
+
+Cannot be specified with the SourceAndType field.
+
+_The `>` syntax is a standard yaml multiline string._
+
+_Until [google/cel-go#203](https://github.com/google/cel-go/issues/203) is
+resolved, dynamic integer fields in CEL must be compared as floats._
+
+```yaml
+spec:
+  filter:
+    cel:
+      parseData: true
+      expression: >
+        ce.type == "dev.knative.observation" &&
+        data.latency > 300.0
```

Comment:
if the express was something like:
```
ce.contenttype == "application/json" &&
data.foo = "bar"
```
but the content type wasn't actually json for a particular event, will the CEL just return 'false' or will it generate an error because it tried to parse the data as json by mistake? ie. are the expressions only evaluated as needed?

Reply:
If `data` were not supplied, CEL would do one of three things:

1. If `ce.contenttype == 'application/json'` is `false`, then short-circuit return `false`.
1. If `ce.contenttype == 'application/json'` is `true`, and `data` was not supplied, CEL would return an error and the filter would need to decide whether errors are fail-closed or fail-open.
1. If the order of the branches were reversed and `ce.contenttype == ...` returns `false`, the result would still be `false`.

There's an outline of the truth-table for how CEL handles errors and unknown values in conditional branches in the CEL Go [Readme](https://github.com/google/cel-go/blob/master/README.md) if you're curious.